### ISSUE TITLE
pin key automatically from inspire-hep when add some item

### ISSUE THREAD

**Issue by dualer** [retorquere/zotero-better-bibtex#1879]
*Please verify **all** of the below before submitting*:

1. you have picked the right category for the issue in the previous screen.
2. You are on the latest release of Zotero
3. in the Zotero addons screen you can see that you have the latest release of BBT (https://github.com/retorquere/zotero-better-bibtex/releases/latest)
4. you are posting a single bug or feature request.
5. the issue has a subject that succinctly describes the problem or question.
6. you are available for follow-up questions and testing.
7. you have included a support log ID here generated by reproducing the problem, selecting the problematic reference(s), right-clicking, and submitting a BBT support log from that popup menu, **not** from the `Help` menu.
8. you have included the *actual output* you got from exporting the items you sent in the debug log, not simplified fictional output or a description of the output, and a sample (*not* a description) of how you want it to look instead, again based on the actual items you selected in the previous step.

Picking the right issue category is really important. Each category (`Export`, `General error`, `Import`, `Key generation`, `Question`) has different instructions for gathering the data necessary required to resolve the issue you are experiencing

The support log is important; it gives me your current BBT settings and a copy of the problematic reference as a test case so I can best replicate your problem. Without it, I'm effectively blind.

**Expected behavior:**

I appreciate this project very much, especially `pin key from inspire-hep` option. If the following function can be added, the world will be better:

 When a user adds some item into `Zotero`, the BBT can pin the key from inspire-hep automatically.

Now, when I add some item into Zotero, BBT can generate the key following ` citation key format`, but it could be done better as I have said above.

**Comment by retorquere** [comment]
A build will drop here in about 10 minutes that adds `[inspire-hep]` as a citekey pattern option. If you combine that with auto-pinning, you would get auto-pinned inspire-hep keys. You may want to do something like `[inspire-hep] | [auth][year]` so items that can't be found on inspire-hep get a sensible fallback key.

**Comment by dualer** [comment]
@retorquere Ok, It works. If the key is pinned from `inpsire-hep` automatically,  could the `pin icon`  be also added automatically?

**Comment by retorquere** [comment]
I think you'll get what you want when you enable [auto-pinning](https://retorque.re/zotero-better-bibtex/installation/preferences/hidden-preferences/#autopindelay). `[inspire-hep]` just fetches the inspireHEP key; pinning is a more general feature where you pin the key to a value by registering it in the `extra` field in a `Citaion Key:` line; if that is present, no matter how it gets there, that will be the key used, and will show the pin icon.

**Comment by dualer** [comment]
Oh, you should know I just want to pin the key from `inspire-hep`. Is it possible to achieve it? This may be not very important, if this is not very easy to achieve, it's also fine now.

**Comment by retorquere** [comment]
You mean you want to auto-pin inspire-hep keys, but not others? I'm can't make a special case of inspire-hep key generation, that path is going to lead to more special-case generation, and I don't want to go down that path.

I have something that could get you close, and it would look something like `[inspire-hep][pin][0] | [auth][year]`, but I'm still conflicted about what I should do about duplicates. I think you'd prefer to have them both get the exact inspire-hep key, but the key generation code will automatically postfix duplicates, so if you import an article that generates key `CMS:2021aly` twice, the 2nd import would get `CMS:2021aly-1` pinned. Really not keen on exempting specific patterns from postfix generation.

**Comment by dualer** [comment]
Yes, this is what I mean. I think this is not only a challenge, but also a chance. Some functions can be added, e.g. when duplication happens, BBT will report the warning, then users can choose to drop the duplication or delete the original item or other manipulation. But the best solution may be that BBT allows duplication when some options like `[allow copy]` are added, so the option is like `[inspire-hep][pin][0][allow copy] | [auth][year]`.

**Comment by retorquere** [comment]
I don't see that as adding much value to be honest. Certainly I'm not adding something that requires UI interaction into the key generation process.

I can add a pinning function like I outlined before, but that will also pin the disambiguator. Not doing so would just lead to duplicate keys in your database, and I don't see how that could ever be desirable.

**Comment by dualer** [comment]
Maybe you can let BBT just don't pin the duplication. The code idea is that we just want to pin the right  key, if not, we woud rather don't pin it.

**Comment by retorquere** [comment]
I don't think that this is a workflow that will be intuitive to understand.

`[inspire-hep][pin]` is as far as I will go (the `[0]` was just so the disambiguator was easy to see, it does nothing special here). It keeps in line with the behavior of the existing general key generation facility. As I said, I don't see value in a workflow where BBT will end up pinning duplicate keys, even if from some perspective these duplicates are the "right keys". If you have a bib file with duplicate keys you will get undesirable behavior during document rendering anyway; with the proposed don't-pin-the-disambiguator process you would have to take care to not import duplicates, but if you do that, `[inspire-hep][pin]` would do the same as the more complicated procedure where BBT would only sometimes pin a key, or only sometimes pin a key without a disambiguator.

`[pin]` would be compatible with auto-pinning; `[pin]` would take precedence. I can see value in this (and auto-pinning) because that is currently the only easy way to get your citekeys into a service like overleaf.

**Comment by retorquere** [comment]
BTW you can always select your whole library and choose "pin from inspireHEP". That will pin the key as-is from inspireHEP, duplicates and all. The pattern generator facility is set up to not generate duplicates, and I don't want to add special exception cases to it.

**Comment by dualer** [comment]
Ok, in fact, users can use `Duplicate Items` to delete duplicates, but we should take care to choose the item having "right" key insead of "false" one. Maybe there are some addons that can throw a warning when users add some duplicate item. As I said, I just don't want to pin some "false" key, e.g. compared to `CMS:2021aly`, `CMS:2021aly-1` is not right.

**Comment by retorquere** [comment]
Which is why I think that selecting and pinning is a better solution for your use-case. BBT citekey generation generates unique keys, by design. Manual pinning van override that.

When Zotero gets a format citation key field, picking the right key on de-duplicate will be easier, but I have no idea about the timeline for that. It's been talked about for years now.

**Comment by dualer** [comment]
I just don't understand is that: in zotero main interface, we can pin the duplicate key using BBT, but pattern generator cannot do it.

**Comment by retorquere** [comment]
That is correct. One is automatic and generates keys without intervention and should not generate duplicate keys because of that. Manually, I'm not standing in your way if you want to have duplicate keys.

**Comment by dualer** [comment]
Got it. I think even if you add pin function, I can't dare to use it, because I don't want to get false key, and I'd rather pin key manually.

**Comment by retorquere** [comment]
TBH I'm hesitant to even add the pin function. The setup of the key generator would have to allow for some tricky workarounds, and it could easily lead to race conditions.

**Comment by dualer** [comment]
You can add it, but as i said, I probably don't use it. You should consider more than me, so do what you want.

**Comment by retorquere** [comment]
Then I won't. Race conditions are very hard to debug if the occur.

**Comment by dualer** [comment]
Ok, that's all right. Maybe there are some better solutions, but I have no ideas now because of my ignorance of underlying structure of BBT. This function is undoubtedly useful, but how to achieve it properly needs more consideration. Thanks for your reply.

**Comment by dualer** [comment]
Oh, I think 

> 🤖 this is your friendly neighborhood build bot announcing [test build 5.4.28.1268](https://github.com/retorquere/zotero-better-bibtex/releases/download/builds/zotero-better-bibtex-5.4.28.1268.xpi) ("inspire-hep pattern")
> 
> Install in Zotero by downloading [test build 5.4.28.1268](https://github.com/retorquere/zotero-better-bibtex/releases/download/builds/zotero-better-bibtex-5.4.28.1268.xpi), opening the Zotero "Tools" menu, selecting "Add-ons", open the gear menu in the top right, and select "Install Add-on From File...".

@retorquere The link is invalid, can you provide another link?

**Comment by retorquere** [comment]
Builds expire in about a week so that if I have made fixes in the interim are not confounding factors. A new build will drop here in 10 minutes or so.

**Comment by dualer** [comment]
No, it doesn't work, the problem related to https://github.com/retorquere/zotero-better-bibtex/issues/1909.

**Comment by retorquere** [comment]
"related to" in what way?

**Comment by dualer** [comment]
> "related to" in what way?

I mean the setting page stay "Better BibTeX is loading".
